mllib/src/main/scala/org/apache/spark/ml/feature/Selector.scala

```diff
@@ -0,0 +1,391 @@
+  }
+
+  @Since("3.1.0")
+  override def copy(extra: ParamMap): Selector[T] = defaultCopy(extra)
+}
+
+/**
+ * Model fitted by [[Selector]].
+ */
+@Since("3.1.0")
+private[ml] abstract class SelectorModel[T <: SelectorModel[T]] (
+    @Since("3.1.0") val uid: String,
+    @Since("3.1.0") val selectedFeatures: Array[Int])
+  extends Model[T] with SelectorParams with MLWritable {
+  self: T =>
+
+  if (selectedFeatures.length >= 2) {
+    require(selectedFeatures.sliding(2).forall(l => l(0) < l(1)),
+      "Index should be strictly increasing.")
+  }
+
+  /** @group setParam */
+  @Since("3.1.0")
+  def setFeaturesCol(value: String): this.type = set(featuresCol, value)
+
+  /** @group setParam */
+  @Since("3.1.0")
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  @Since("3.1.0")
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val outputSchema = transformSchema(dataset.schema, logging = true)
+
+    SelectorModel.transform(dataset, selectedFeatures, outputSchema, $(outputCol), $(featuresCol))
+  }
+
+  @Since("3.1.0")
+  override def transformSchema(schema: StructType): StructType = {
+    SchemaUtils.checkColumnType(schema, $(featuresCol), new VectorUDT)
+    val newField =
+      SelectorModel.prepOutputField(schema, selectedFeatures, $(outputCol), $(featuresCol))
+    SchemaUtils.appendColumn(schema, newField)
+  }
+}
+
+object SelectorModel {
+
+  private[ml] def transform(
+      dataset: Dataset[_],
+      selectedFeatures: Array[Int],
+      outputSchema: StructType,
+      outputCol: String,
+      featuresCol: String): DataFrame = {
+    val newSize = selectedFeatures.length
+    val func = { vector: Vector =>
+      vector match {
+        case SparseVector(_, indices, values) =>
+          val (newIndices, newValues) =
+            compressSparse(indices, values, selectedFeatures)
+          Vectors.sparse(newSize, newIndices, newValues)
+        case DenseVector(values) =>
+          Vectors.dense(selectedFeatures.map(values))
+        case other =>
+          throw new UnsupportedOperationException(
+            s"Only sparse and dense vectors are supported but got ${other.getClass}.")
+      }
+    }
+
+    val transformer = udf(func)
+    dataset.withColumn(outputCol, transformer(col(featuresCol)),
+      outputSchema(outputCol).metadata)
+  }
+
+  /**
+   * Prepare the output column field, including per-feature metadata.
+   */
+  private[ml] def prepOutputField(
+      schema: StructType,
+      selectedFeatures: Array[Int],
+      outputCol: String,
+      featuresCol: String): StructField = {
+    val selector = selectedFeatures.toSet
+    val origAttrGroup = AttributeGroup.fromStructField(schema(featuresCol))
+    val featureAttributes: Array[Attribute] = if (origAttrGroup.attributes.nonEmpty) {
+      origAttrGroup.attributes.get.zipWithIndex.filter(x => selector.contains(x._2)).map(_._1)
+    } else {
+      Array.fill[Attribute](selector.size)(NominalAttribute.defaultAttr)
```

Comment:
I guess we need to change this place:
```
Array.fill[Attribute](selector.size)(NominalAttribute.defaultAttr)
```
this comes from `ChiSqSelector`, which treat each column as `Nominal`.
Since `selector`s now support numerical columns, like `FValue`/`ANOVA`, so we may need change it to:
1, only attach output vector length by method `SchemaUtils.updateAttributeGroupSize`, you can refer to the it usage in GMM;
2, add a new method to let each subclass tell the features's type